### PR TITLE
Add CHPL_HOST_MACHINE, CHPL_TARGET_MACHINE

### PR DESCRIPTION
**Pull request by mppf**
This PR resolves two related problems:
 1. Compiler builds were not differentiated based on architecture (#11386)
 2. `native` as an architecture is more ambiguous than we would like (#9247)

How do paths change? At a very high level:

_ | Before this PR | After this PR
-- | ------------ | -------------
bin/.../chpl | bin/linux64/chpl | bin/linux64-x86_64/chpl
third-party/*/install/.../ | linux64-gnu-native-none | linux64-x86_64-native-gnu-none

Note -none above is about CHPL_LIBPIC.

Further examples for third-party/*/install/.../ :
 
_ | 3p config dir
-- | ------------
CHPL_TARGET_ARCH=native | linux64-x86_64-native-clang-none
CHPL_TARGET_ARCH=none | llinux64-x86_64-none-clang-none
CHPL_TARGET_ARCH=unknown | linux64-x86_64-native-clang-none
CHPL_TARGET_ARCH=sandybridge | linux64-x86_64-sandybridge-clang-none

Summary of Discussion:
 * Originally, we talked about including general information (e.g. `x86_64`) in CHPL_TARGET_ARCH. But, we wanted to keep printchplenv showing environment variable values unmodified, but also wanted the machine type to be included in both `CHPL_TARGET_ARCH=native` and `CHPL_TARGET_ARCH=sandybridge` scenarios, so instead of prepending/appending machine type to CHPL_TARGET_ARCH, the next approach added a new variable CHPL_TARGET_MACHINE
 * the bin/ directory is e.g. linux64-x86_64. Doesn't currently include HOST_ARCH since we don't (currently) specialize the compiler build. The order linux64-x86_64 matches what we currently do for third-party install dirs. 
 * Even if a specialization is selected we could include the general architecture, e.g. `x86_64-sandybridge`
 * `x86_64-none` should shorten to just `x86_64`
 * default CHPL_HOST_CPU should be `none`
 * In later discussion we realized that CHPL_TARGET_ARCH is a better name for e.g. x86_64 than CHPL_TARGET_MACHINE, and CHPL_TARGET_CPU is a better name for what we were doing with CHPL_TARGET_ARCH. This portion is under discussion in issue #11740.

Detailed list of changes in this PR:
 * updates chpl_arch.py to include a new function, get_default_machine, which defaults to the `uname -m` for arch=none/native/unknown and computes the machine type from more detailed arch settings.
 * adds chpl_machine.py which outputs the selected machine type
 * updates printchplenv to output `CHPL_HOST_ARCH`, including for compiler build paths. Note that this is not output by default and defaults to "none" (indicating that we don't specialize the compiler build).
 * updates printchplenv to output `CHPL_TARGET_MACHINE` and `CHPL_HOST_MACHINE`. These store a machine type (e.g. x86_64).
 * adds a new chplconfig script `chpl_bin_subdir.py` to get the bin directory part (since it's not just platform anymore and we use it in a lot of places such as setchplenv scripts). This normally outputs `linux64-x86_64` but if specialization is activated (which normally happens only for --target which is currently unused) it can be e.g. `linux64-x86_64-native`.
 * updates printchplenv to have internal variables CHPL_HOST_BIN_SUBDIR and CHPL_TARGET_BIN_SUBDIR computed by above script
 * updates third_party_utils default_uniq_cfg_path to include the result of chpl_arch.get
 * updates compiler build to use include CHPL_MAKE_HOST_BIN_SUBDIR
 * updates setchplenv.* scripts for the new bin/ path
 * adds a script named `chpl` at the old bin path to print a warning and run the compiler

- [x] checked Cray module builds
- [x] update lists of printchplenv vars, including in overrides.py
- [x] full local testing

Reviewed by @ronawho and @ben-albrecht - thanks!

### CODE REVIEW

**Comment by mppf** [comment]
@dmk42 @ronawho @gbtitus - I'm interested in your feedback on this prototype to resolve some of the "ARCH" issues. Thanks!

**Comment by ronawho** [comment]
Overall, I like this approach: 

>How should we write a disambiguated architecture? I'm advocating for something along the lines of native-x86_64 for example. A reasonable alternative would be to go directly to haswell etc. but these are both less likely to be understood by a user and also might leave out optimization opportunities (I think backend compiler -march=native can be more aggressive). Anyway, we could also consider x86_64-native or native@x86_64 or native(x86_64) or any number of things.

I like native-x86_64

> What can we call the disambiguated architecture? I called the tuple field narch (for "normalized architecture") but that's almost certainly not the best name...

I'd be ok with either "normalized architecture" or "disambiguated architecture". Maybe shortened to n_arch/norm_arch or dis_arch in the python scripts.

> How should we update the bin directory path? Right now it makes e.g. bin/native-x86_64-linux64 but it could be bin/native-x86_64/linux64 or bin/linux64-native-x86_64 or use different separators or any number of possibilities. I started with the "arch" first because that is what GNU target triples do (https://wiki.osdev.org/Target_Triplet)

For third party dirs we do `platform-compiler-architecture` -- so my preference would be to keep it similiar and do `bin/linux64-x86_64`. Actually why isn't host-compiler included there too? Should that be `bin/linux64-gnu-x86_64`? Shouldn't building with say host_comp=gnu and host_comp=clang result in 2 distinct builds?

> In how much detail should we elaborate on native or none or unknown ? I just used get_native_machine() which basically is the same as uname -p for now. However, we could consider getting much more detail, either from /proc/cpuinfo or from gcc -march=native -E -v - </dev/null 2>&1. I personally think simply listing the very general architecture information in these cases is sufficient. Additionally, I'd expect that we could choose to change this behavior in the future without causing too much trouble.

I'm happy with `get_native_machine()`

**Comment by mppf** [comment]
@ronawho - thanks for your feedback.

> For third party dirs we do platform-compiler-architecture -- so my preference would be to keep it similiar and do bin/linux64-x86_64. Actually why isn't host-compiler included there too? Should that be bin/linux64-gnu-x86_64? Shouldn't building with say host_comp=gnu and host_comp=clang result in 2 distinct builds?

To be clear, wouldn't the typical case be `bin/linux64-native-x86_64` with that proposal, because the "host architecture" is native-x86_64 rather than just x86_64?

I don't think we need to include compiler for executable programs since there isn't an issue of binary (in)compatibility when linking with other programs. If the compiler one day presents functionality as a .a or .so, that would be in a lib/ directory and would need the compiler to form part of the path.

We might decide that including the host compiler is worthwhile for some other reason (easier testing? user convenience?) but I don't see a compelling use case for it. That said, it wouldn't bother me if we did include it in the bin/ subdirectory.

**Comment by ronawho** [comment]
>>For third party dirs we do platform-compiler-architecture -- so my preference would be to keep it similiar and do bin/linux64-x86_64. Actually why isn't host-compiler included there too? Should that be bin/linux64-gnu-x86_64? Shouldn't building with say host_comp=gnu and host_comp=clang result in 2 distinct builds?

>To be clear, wouldn't the typical case be bin/linux64-native-x86_64 with that proposal, because the "host architecture" is native-x86_64 rather than just x86_64?

I don't think we should specialize the compiler build. i.e. we don't throw -march=native for compiler builds and I don't think we want to. If anything the full host architecture would be `none-x86_64`, but I think it should always be none (no specialization), so I think we could elide that part.

>I don't think we need to include compiler for executable programs since there isn't an issue of binary (in)compatibility when linking with other programs. If the compiler one day presents functionality as a .a or .so, that would be in a lib/ directory and would need the compiler to form part of the path.

Do we have to be concerned about binary compatibility between the compiler and llvm? What would happen if a user built our compiler with gnu and llvm with intel?

>We might decide that including the host compiler is worthwhile for some other reason (easier testing? user convenience?) but I don't see a compelling use case for it. That said, it wouldn't bother me if we did include it in the bin/ subdirectory.

I have a slight preference for including it.

**Comment by ronawho** [comment]
Is there anything we have to do for launchers and/or llvm here?

**Comment by mppf** [comment]
> I don't think we should specialize the compiler build. i.e. we don't throw -march=native for compiler builds and I don't think we want to. If anything the full host architecture would be none-x86_64, but I think it should always be none (no specialization), so I think we could elide that part.

I can see this working in one of two ways:
 1. While we might have native-x86_64, instead of none-x86_64 we'd simply call it x86_64.
 2. We add another (probably hidden) variable to printchplenv that represents the result of get_native_machine and use that in certain places in Makefiles.

I think the first sounds better here but I'm interested in your viewpoint. With that solution, we'd want to make the default CHPL_HOST_ARCH be `none`.

> Do we have to be concerned about binary compatibility between the compiler and llvm? What would happen if a user built our compiler with gnu and llvm with intel?

* If it's the LLVM we include, I expect we'll include the compiler in the third-party/llvm/install subdirectory (at least we do that now). That means that you couldn't have `chpl` built for intel and GCC at the same time, but it would refer to whatever LLVM was appropriate and you used last.
* If it's the system LLVM, we have absolutely no control over what compiler built it. In fact I don't think we can even know what compiler built it. I believe that the normal situation in linux distributions is to require libraries have compatability with the system-wide-default GCC (including the accompanying libstdc++).
* So far, our recommendation when building the compiler is "use GCC". I have trouble coming up with a use case where using the Intel compiler to build `chpl` would be useful, but you never know.

> Is there anything we have to do for launchers and/or llvm here?

I'm sure that there is, but I wanted to discuss before trying to do the rest of the implementation.

**Comment by gbtitus** [comment]
This looks generally good to me.

I have a slight preference for e.g. x86_64-native because I view none/native/unknown as subsetting the named architecture, but if overruled I will not weep.

> Is there anything we have to do for launchers and/or llvm here?

It seems like for the launcher we need to do whatever we do for the runtime, except using host arch info rather than target arch info.

**Comment by mppf** [comment]
> I have a slight preference for e.g. x86_64-native because I view none/native/unknown as subsetting the named architecture, but if overruled I will not weep.

I wonder, would you also want to have e.g. x86_64-sandybridge ? That is, even when the selected arch isn't "native", "none", or "unknown", we could put the general architecture in front of it? The only reason to do that would be some kind of consistency / something perhaps easier to understand as a user. We'd still let users set CHPL_TARGET_ARCH=sandybridge e.g..

**Comment by gbtitus** [comment]
> I wonder, would you also want to have e.g. x86_64-sandybridge ?  ...

For consistency, yes, I think that probably is what we'd want.

**Comment by mppf** [comment]
If we were doing e.g. `x86_64-sandybridge` then I'd agree that `x86_64` should come first as in `x86_64-native`. The drawback of this is that what the user specified will generally appear after the general architecture.

**Comment by gbtitus** [comment]
> ...  The drawback of this is that what the user specified will generally appear after the general architecture.

Yes, that's a bit unfortunate.  And although at present that specification is done only on Cray X* systems and indirectly, by having a `craype-*` module loaded, either or both of those limitations might not be true in the future.  This gets back to something I tried to express a few months ago when we were discussing this in email, that our `CHPL_TARGET_ARCH` setting was confusing because sometimes it was indeed an architecture (`x86_64`, for example), but sometimes it was an implementation of an architecture (`sandybridge`, for example).  It would be nice to clean that aspect up in this work.

**Comment by ronawho** [comment]
>> I wonder, would you also want to have e.g. x86_64-sandybridge ? ...

> For consistency, yes, I think that probably is what we'd want.

Hmm, I find it a little odd that if a user sets `CHPL_TARGET_ARCH=sandybridge` then printchplenv will print someting other than what they set. Although, I guess that's what we're propsing for none/native/unknown already. 

This makes me wonder if we should only encode the disambiguation part (x86_64) into the compiler/runtime/launcher paths, but not actually modify the value of `CHPL_{HOST,TARGET}_ARCH`. 

Like have `CHPL_TARGET_ARCH` remain the same, but then add a `CHPL_DISAMBIGUATED_TARGET_ARCH` or something?

**Comment by mppf** [comment]
@ronawho - yeah, I think we could independently (from the other choices) decide whether printchplenv would show the additional information. However I'd personally favor having printchplenv show the additional information for the sake of being clear with the user about what was inferred. I think the main use case for `printchplenv` for users is to understand what the settings are, and if this is an important setting that is inferred, that's useful to show. For example, if I set CHPL_TARGET_ARCH=native, I'd be happy to see `printchplenv` show the architecture (e.g. `x86_64`) in addition. I think the challenge there though is how to be clear about which part is inferred and which part is not. Using a different variable is one way to do that. Another way is having a predictable and intuitive way to combine the specified part with the inferred part, which I think `native-x86_64` does a reasonable job of, and `x86_64-native` does less well but would still be acceptable I think.

**Comment by ronawho** [comment]
Today things in chplenv are either fully inferred or we just take what the user gave us, so overriding what a user explicitly said with a partially inferred value seems wrong to me. I'd be in favor of something like `CHPL_DISAMBIGUATED_TARGET_ARCH`

**Comment by mppf** [comment]
@ronawho - if we were adding a new (user-visible) `printchplenv` variable anyway, should we just add a new variable for the machine type / general architecture? E.g. on that stores `x86_64` or `aarch64` etc ? I think the main challenge I have with that is coming up with a reasonable name. ("Architecture" seems most reasonable but we are already using that).

**Comment by dmk42** [comment]
I use printchplenv to see what I set, so I am also in favor of having a separate variable to provide the additional information, rather than reporting a variable I set with a value that is not what I set.

Something like CHPL_HOST_MACHINE_TYPE might be reasonable.  That fits with the fact that we're using the equivalent of `uname -m` (not `uname -p`).

**Comment by mppf** [comment]
@ronawho - I've updated this PR with a new suggestion based on the discussion. The PR message summarizes but of course you can look at the diff. What do you think?

File: util/chplenv/chpl_arch.py [hunk 1]
```diff
@@ -550,6 +549,23 @@ def get(location, map_to_compiler=False, get_lcd=False):
...
+#  * otherwise returns the machine type corresponding to the selected arch
+@memoize
+def get_default_machine(location):
+    (flag, arch) = get(location)
+
+    # Native/none/unknown just get the native machine by default
+    if arch == "native" or arch == "none" or arch == "unknown":
+      return get_native_machine()
+    # Otherwise, compute the machine based on the architecture selected
+    if is_known_arm(arch):
```
Comment (ronawho) [inline util/chplenv/chpl_arch.py]: This seems likely to easily become out of date.

File: util/chplenv/chpl_arch.py [hunk 2]
```diff
@@ -571,8 +587,9 @@ def _main():
 
     if options.compflag:
         stdout.write("{0}=".format(flag))
-    stdout.write("{0}\n".format(arch))
-
+        stdout.write("{0}\n".format(arch))
+    else:
+        stdout.write("{0}\n".format(arch))
```
Comment (ronawho) [inline util/chplenv/chpl_arch.py]: What's the motivation for this change?

File: util/chplenv/chpl_arch.py (hunk 1 above)
Comment (mppf) [reply util/chplenv/chpl_arch.py]: I agree. I think we could do something better using what we already have in `feature_sets` but I thought that'd be a task for some one with more Python skills than me. I view this version as an OK 1st solution, though.

File: util/chplenv/chpl_bin_subdir.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python
+
+import optparse
+import os
+from sys import stderr, stdout
```
Comment (ronawho) [inline util/chplenv/chpl_bin_subdir.py]: There is only one use of stdout here, so I'd prefer to just stick with fully qualified names

File: util/chplenv/chpl_arch.py [hunk 4]
```diff
@@ -436,23 +436,22 @@ def get_module_lcd_arch(platform_val, arch):
     else:
         return 'none'
 
-# get_lcd has no affect on non cray systems and is intended to be used to get
+# get_lcd has no effect on non cray systems and is intended to be used to get
 # the correct runtime and gen directory.
-@memoize
```
[the remark this answers is not in the document]
Comment (mppf) [reply util/chplenv/chpl_arch.py]: Didn't mean to, will fix

File: util/chplenv/chpl_arch.py (hunk 2 above)
Comment (mppf) [reply util/chplenv/chpl_arch.py]: Mattered in earlier version, thought I undid that, will fix

File: util/chplenv/chpl_bin_subdir.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    # platform
+    result = platform
+    # compiler -- included for target
+    if location == 'target':
+      result += '-' + compiler
+    # machine
+    result += '-' + machine
+    # arch
+    if arch != 'none' and arch != 'unknown':
```
Comment (ronawho) [inline util/chplenv/chpl_bin_subdir.py]: Why not include arch here?

File: util/chplenv/chpl_bin_subdir.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    compiler = chpl_compiler.get(location)
+    machine = chpl_machine.get(location)
+    (flag, arch) = chpl_arch.get(location, map_to_compiler=True,
+                                 get_lcd=chpl_home_utils.using_chapel_module())
+
+    # platform
+    result = platform
+    # compiler -- included for target
+    if location == 'target':
+      result += '-' + compiler
```
Comment (ronawho) [inline util/chplenv/chpl_bin_subdir.py]: I still have a slight preference for including compiler for host portions

File: util/chplenv/chpl_bin_subdir.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+chplenv_dir = os.path.dirname(__file__)
+sys.path.insert(0, os.path.abspath(chplenv_dir))
+
+import chpl_home_utils
+import chpl_platform, chpl_compiler, chpl_machine, chpl_arch
+
+from utils import memoize
+
+@memoize
+def get(location):
```
Comment (ronawho) [inline util/chplenv/chpl_bin_subdir.py]: I find location to be a confusing name. Most chplenv scripts just call this "flag", which is relatively meaningless, but I don't find that confusing. location just doesn't seem like an accurate description to me.

File: util/chplenv/chpl_machine.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python
+import collections
```
Comment (ronawho) [inline util/chplenv/chpl_machine.py]: unused import, so is platform the string one. I'd also prefer to keep stderr and stdout as fully qualified. 

I think you're also missing an import for `error()`

File: util/chplenv/chpl_bin_subdir.py (hunk 6 above)
Comment (mppf) [reply util/chplenv/chpl_bin_subdir.py]: We talked about having it be `bin/linux64-x86_64/chpl`.. Adding the compiler would make it `bin/linux64-gnu-x86_64/chpl` and I thought we didn't want that. If this is really the path for a *bin* directory it shouldn't matter what compiler was used to make it, since it's not libraries (static or dynamic) but rather standalone programs.

File: util/chplenv/third_party_utils.py [hunk 9]
```diff
@@ -15,12 +15,11 @@
...
 def default_uniq_cfg_path():
-    arch_val = chpl_arch.get('target', map_to_compiler=True,
-                             get_lcd=using_chapel_module()).arch
-    return '{0}-{1}-{2}-{3}'.format(chpl_platform.get('target'),
-                                    chpl_compiler.get('target'),
-                                    arch_val,
-                                    chpl_lib_pic.get())
+    pic = chpl_lib_pic.get()
+    result = chpl_bin_subdir.get('target')
+    if pic != 'none':
```
Comment (ronawho) [inline util/chplenv/third_party_utils.py]: Why change this to not include pic if it's none?

File: util/chplenv/chpl_bin_subdir.py (hunk 6 above)
Comment (mppf) [reply util/chplenv/chpl_bin_subdir.py]: The only reason I put the *bin* directory here was because it matched what we were doing for third-party runtime dirs (so I could move most of the logic in default_uniq_cfg_path to here). But we could handle that separately if we were willing to change the order. Or just not call this from default_uniq_cfg_path.

File: util/setchplenv.bash [hunk 10]
```diff
@@ -38,11 +38,10 @@ fi
...
 
-export CHPL_HOST_PLATFORM=`"$CHPL_HOME"/util/chplenv/chpl_platform.py`
-echo "Setting CHPL_HOST_PLATFORM to $CHPL_HOST_PLATFORM"
+CHPL_BIN_SUBDIR=`"$CHPL_HOME"/util/chplenv/chpl_bin_subdir.py`
 
-export PATH="$CHPL_HOME"/bin/$CHPL_HOST_PLATFORM:"$CHPL_HOME"/util:"$MYPATH"
-echo "Updating PATH to include $CHPL_HOME/bin/$CHPL_HOST_PLATFORM"
+export PATH="$CHPL_HOME"/bin/$CHPL_BIN_SUBDIR:"$CHPL_HOME"/util:"$MYPATH"
+echo "Updating PATH to include $CHPL_HOME/bin/$CHPL_BIN_SUBDIR"
 echo "                     and $CHPL_HOME/util"
```
Comment (ronawho) [inline util/setchplenv.bash]: Do we have to change non bash ones too?

File: util/chplenv/chpl_bin_subdir.py (hunk 7 above)
Comment (mppf) [reply util/chplenv/chpl_bin_subdir.py]: I copied it from chpl_arch. I'm happy to change it to flag in both places.

File: util/chplenv/third_party_utils.py (hunk 9 above)
Comment (mppf) [reply util/chplenv/third_party_utils.py]: It makes sense to me to omit it if it's none but we don't have to keep that change.

File: util/setchplenv.bash (hunk 10 above)
Comment (mppf) [reply util/setchplenv.bash]: Yes but I wanted to get feedback on the approach before changing *everything*

**Comment by ronawho** [comment]
I made a lot of mostly nitpicky/style comments, but I think I'm on board with the overall change. 

I find the summary to be pretty hard to follow. I think it'd be helpful to give an example of what the unique paths looked like before vs now for the compiler (host) and third-party/runtime (target) dirs.

File: util/chplenv/third_party_utils.py [hunk 11]
```diff
@@ -15,12 +15,11 @@
...
 @memoize
 def default_uniq_cfg_path():
-    arch_val = chpl_arch.get('target', map_to_compiler=True,
-                             get_lcd=using_chapel_module()).arch
-    return '{0}-{1}-{2}-{3}'.format(chpl_platform.get('target'),
-                                    chpl_compiler.get('target'),
-                                    arch_val,
-                                    chpl_lib_pic.get())
+    pic = chpl_lib_pic.get()
+    result = chpl_bin_subdir.get('target')
```
Comment (ronawho) [inline util/chplenv/third_party_utils.py]: Hmm, Calling `bin_subdir` to get the the library_subdir is a little odd. Maybe we can change it to chpl_unique_subdir or something?

Comment (mppf) [reply util/chplenv/third_party_utils.py]: I just stopped doing that. Even if they're similar I think at the moment it's better to not make them call each other.

File: util/chplenv/chpl_bin_subdir.py (hunk 6 above)
Comment (ronawho) [reply util/chplenv/chpl_bin_subdir.py]: Michael convinced me offline to leave compiler out of host portions.

Comment (mppf) [reply util/chplenv/chpl_bin_subdir.py]: My main concern was that people will find the -gnu there confusing and interpret it to mean that the compiler is configured to use GCC as the C backend. Additionally I mentioned that there are other ways to get toggling CHPL_HOST_COMPILER to trigger relinking. (And that appears to work now without any change).

**Comment by ben-albrecht** [comment]
We could ease the transition of user `bashrc` updates by continuing to generate something called `chpl` to the old path, which emits a warning that the path name has been updated.

**Comment by mppf** [comment]
I do think that `CHPL_TARGET_CPU` would be a better name that `CHPL_TARGET_ARCH` the way we are using it. Then the new `CHPL_TARGET_MACHINE` could be `CHPL_TARGET_ARCH`, which I do think is slightly more intuitive for what it is. I'm just uncertain if this is clearer to anybody else and if it's worth the potential for confusion to change the meaning of `CHPL_TARGET_ARCH`.

**Comment by gbtitus** [comment]
> I do think that `CHPL_TARGET_CPU` would be a better name that `CHPL_TARGET_ARCH` the way we are using it. Then the new `CHPL_TARGET_MACHINE` could be `CHPL_TARGET_ARCH`, ...

I agree, assuming I'm understanding you correctly.  I think of "architecture" as the broader term (examples: x86_64, aarch64, m68k), and "cpu" as the more specific one (examples: sandybridge, thunder-x2, m68030).

While writing this it occurred to me that I'm really not that fond of "thunder-x2" because as I understand it that's just a trademark for an implementation of the ARMv8 ISA, and any other implementation of ARMv8 ought to behave the same, functionally.  So it seems like "armv8" might make more sense.  But maybe there are non-ISA differences (instruction timing, say) that would lead to variations in codegen compared to other implementations?  If so then it's okay.  And in any case that bridge is probably already too burnt up to retreat across ...

**Comment by dmk42** [comment]
The Cray module is arm-thunderx2, so that's what we have to work with.  It does have its own idiosyncrasies, and when you translate it to a gcc or clang argument, it becomes the very specific thunderx2t99 with the instruction timings matched exactly to that CPU.

Whether we make a change or not, it will be confusing.  In gcc and clang, x86 targets use `-march` for the specific CPU.  In other targets, they use `-mcpu` for that. If we change, we'll be confusing when the target is x86.  If we don't change, we'll be confusing when the target is other than x86.

The non-x86 usage is more intuitive, though.  I won't advocate for changing, but it won't bother me if everyone else is on board.

**Comment by gbtitus** [comment]
> The Cray module is arm-thunderx2, so that's what we have to work with. It does have its own idiosyncrasies, and when you translate it to a gcc or clang argument, it becomes the very specific thunderx2t99 with the instruction timings matched exactly to that CPU.

Just to make sure I'm thinking about this right, does that `thunderx2t99` gcc/clang target differ from other ARMv8 targets (if there are any?) only in instruction timings, or are there functional differences also?  Or putting it another way, would code generated for any ARMv8 gcc/clang target run on every ARMv8 chipset without fail, though perhaps not at optimal speed?

(Agreed that we want optimal performance on ARM-based Cray compute blades, but I'm trying to think about how we integrate that with our approach to non-Cray ARM targets also.)

**Comment by dmk42** [comment]
ThunderX2 code does not always run on ThunderX, so apparently there are some ISA differences.

**Comment by ben-albrecht** [comment]
I think the redefining of `CHPL_TARGET_ARCH` to mean `x86_64` vs `aarch64` makes a lot of sense.

As for a new name for `CHPL_TARGET_ARCH`, I think `CHPL_TARGET_MICROARCH` is a good contender with `CHPL_TARGET_CPU`. Microarchitecture seems to be consistent with how vendors refer to their specific ISA implementations.

**Comment by mppf** [comment]
I wonder about something like `CHPL_TARGET_ARCH_EXTENSIONS`

**Comment by dmk42** [comment]
I would think that would cause more confusion.  A variable with that name sounds like it should contain a list of things like "sse,sse2,ssse3,sse4.1,sse4.2".

**Comment by ronawho** [comment]
I think`CHPL_TARGET_ARCH` -> `CHPL_TARGET_CPU` needs to be split out into a separate PR and we need to figure out how we're going to document this transition and not make it a huge pain for users. This is a very disruptive change IMO

**Comment by gbtitus** [comment]
> I think`CHPL_TARGET_ARCH` -> `CHPL_TARGET_CPU` needs to be split out into a separate PR and we need to figure out how we're going to document this transition ...

I don't disagree, but I'm having trouble seeing a way to fix this issue itself, dependably, without impacting users just as much.  Maybe introduce a new, hidden, env var based on the `uname -m` output and add that to the compiler, runtime, and third-party unique config paths?

**Comment by gbtitus** [comment]
> Maybe there's isn't a good way to avoid the user impact, but I do not want us to make this change without thinking about the impact on users, and I really really do not want this change lobbed in with another already large-ish and core infrastructure change.

I don't believe we've all had zero thought about the user impact, but it's definitely something we should tackle explicitly.  If nothing else, making and reviewing the documentation changes that would need to be part of this will give us a good handle on that impact.

Just to press a little more, though, do you have a low-impact solution in mind for the current ambiguity problem in the unique config paths?

**Comment by mppf** [comment]
Friends, let's move the discussion of the rename portion to PR #11713. This PR will no longer include the renaming and will instead add CHPL_TARGET_MACHINE and leave CHPL_TARGET_ARCH alone.

**Comment by ronawho** [comment]
>>Maybe there's isn't a good way to avoid the user impact, but I do not want us to make this change without thinking about the impact on users, and I really really do not want this change lobbed in with another already large-ish and core infrastructure change.

>I don't believe we've all had zero thought about the user impact, but it's definitely something we should tackle explicitly. If nothing else, making and reviewing the documentation changes that would need to be part of this will give us a good handle on that impact.

>Just to press a little more, though, do you have a low-impact solution in mind for the current ambiguity problem in the unique config paths?

Just to be clear, I was onboard with adding CHPL_TARGET_MACHINE to resolve the ambiguity problem, it was the renaming that I wasn't thrilled with. 

> Friends, let's move the discussion of the rename portion to PR #11713. This PR will no longer include the renaming and will instead add CHPL_TARGET_MACHINE and leave CHPL_TARGET_ARCH alone. 

👍  thanks

File: util/chplenv/chpl_machine.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python
+import optparse
+import os
+import sys
+
+chplenv_dir = os.path.dirname(__file__)
+sys.path.insert(0, os.path.abspath(chplenv_dir))
```
Comment (ronawho) [inline util/chplenv/chpl_machine.py]: Sorry, https://github.com/chapel-lang/chapel/pull/11725 changed things out from under you. Please remove the path hacking (and os/sys includes if they're not needed elsewhere.) Ditto for chpl_bin_subdir.py 

printchplenv is also now a bash wrapper, so you'll have to update printchplenv.py instead

File: util/chplenv/__init__.py [hunk 13]
```diff
@@ -14,6 +15,7 @@
     'chpl_launcher',
     'chpl_llvm',
     'chpl_locale_model',
+    'chpl_machine',
```
Comment (ronawho) [inline util/chplenv/__init__.py]: __init__ was removed in https://github.com/chapel-lang/chapel/pull/11725, please add these to chplenv.py instead

File: test/chplenv/autocomplete/autocomplete.chpl [hunk 14]
```diff
@@ -10,12 +10,6 @@ use Spawn;
 
 var home = CHPL_HOME;
 
-extern proc getenv(name: c_string): c_string;
-extern proc setenv(name: c_string, val: c_string, overwrite: c_int): c_int;
-
-var path = getenv("PATH"):string;
-setenv("PATH", (home + "/bin/" + CHPL_HOST_PLATFORM + ":" + path).c_str(), 1);
-
```
Comment (ronawho) [inline test/chplenv/autocomplete/autocomplete.chpl]: Why is this no longer needed?

Comment (mppf) [reply test/chplenv/autocomplete/autocomplete.chpl]: See commit ed0884bac20bee1aadeba08d8869bab9e4eaf8d0 -- I updated the script it calls, which seems a more reasonable place to handle the case of PATH not being set to include chpl.

**Comment by mppf** [comment]
@ronawho - I think this is ready to go (but I'm not going to merge it today). Please let me know if you have more review feedback.

File: util/chplenv/chpl_bin_subdir.py (hunk 5 above)
Comment (mppf) [reply util/chplenv/chpl_bin_subdir.py]: It's my response to your suggestion:
https://github.com/chapel-lang/chapel/pull/11638#issuecomment-438765716

if the arch is none or unknown, we elide it from the bin/ path. Note that we're not planning to specialize the compiler build, so we need some rule to keep the bin/ subdirectory simple.

File: test/mason/EXECENV [hunk 15]
```diff
@@ -2,7 +2,8 @@
 
 import os
 import sys
+import subprocess
```
Comment (ronawho) [inline test/mason/EXECENV]: subprocess import is not needed

File: util/chplenv/chpl_arch.py [hunk 16]
```diff
@@ -477,62 +460,134 @@ def get(location, map_to_compiler=False, get_lcd=False):
...
     elif 'cray' in compiler_val:
-        return arch_tuple('none', 'none')
+        return 'none'
     elif 'ibm' in compiler_val:
-        return arch_tuple('none', 'none')
+        return 'none'
+
+    return arch
+
+def default_arch(flag):
```
Comment (ronawho) [inline util/chplenv/chpl_arch.py]: Memoize this and other new functions?

**Comment by ronawho** [comment]
Generally this looks great to me. I need to spend some more time reviewing chpl_arch.py, but I'd like to do that with fresh eyes on Monday

File: util/chplenv/chpl_arch.py (hunk 16 above)
Comment (mppf) [reply util/chplenv/chpl_arch.py]: done

File: test/mason/EXECENV (hunk 15 above)
Comment (mppf) [reply test/mason/EXECENV]: I changed it to a bash script and many to links. Didn't see problems until I paratested (which is apparently the environment that leaves chpl not in PATH).

File: util/chplenv/chpl_arch.py [hunk 17]
```diff
@@ -477,62 +461,136 @@ def get(location, map_to_compiler=False, get_lcd=False):
...
+            arch = 'native'
+
+    return arch
+
+# Returns the machine type for the passed arch.
+#  * if arch is native/none/unknown, return result of get_native_machine
+#  * otherwise returns the machine type corresponding to the selected arch
+@memoize
+def machine_for_arch(arch):
+    if arch == "native" or arch == "none" or arch == "unknown":
```
Comment (ronawho) [inline util/chplenv/chpl_arch.py]: Looks like there's inconsistent indentation here and other places. Can you unify to 4 spaces per indention level?

File: util/chplenv/chpl_arch.py [hunk 18]
```diff
@@ -546,13 +604,22 @@ def get(location, map_to_compiler=False, get_lcd=False):
     return arch_tuple(flag or 'none', arch or 'unknown')
 
 
+# Returns the default machine.
+#  * if arch is native/none/unknown, return result of get_native_machine
+#  * otherwise returns the machine type corresponding to the selected arch
+@memoize
+def get_default_machine(flag):
+    (flag, arch) = get(flag)
```
Comment (ronawho) [inline util/chplenv/chpl_arch.py]: `(flag, arch) = get(flag)` => `(_, arch) = get(flag)`?

File: util/chplenv/chpl_bin_subdir.py [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+    # compiler is never included since it shouldn't be needed
+    # for a bin/ path.
+
+    # machine
+    result += '-' + machine
+
+    # arch
+    if arch != 'none' and arch != 'unknown':
+      result += '-' + arch
```
Comment (ronawho) [inline util/chplenv/chpl_bin_subdir.py]: Indention level again
